wallet/txauthor/author.go

```diff
@@ -0,0 +1,200 @@
+		remainingAmount := inputAmount - targetAmount
+		if remainingAmount < maxRequiredFee {
+			targetFee = maxRequiredFee
+			continue
+		}
+
+		unsignedTransaction := &wire.MsgTx{
+			Version:  wire.TxVersion,
+			TxIn:     inputs,
+			TxOut:    outputs,
+			LockTime: 0,
+		}
+		changeIndex := -1
+		changeAmount := inputAmount - targetAmount - maxRequiredFee
+		if !txrules.IsDust(changeAmount, txsizes.P2PKHPkScriptSize, relayFeePerKb) {
+			changeScript, err := fetchChange()
+			if err != nil {
+				return nil, err
+			}
+			if len(changeScript) > txsizes.P2PKHPkScriptSize {
+				return nil, errors.New("fee estimation requires change " +
+					"scripts no larger than P2PKH output scripts")
+			}
+			change := wire.NewTxOut(int64(changeAmount), changeScript)
+			l := len(outputs)
+			unsignedTransaction.TxOut = append(outputs[:l:l], change)
+			changeIndex = l
+		}
+
+		return &AuthoredTx{
+			unsignedTransaction,
+			scripts,
+			inputAmount,
+			changeIndex,
+		}, nil
+	}
+}
+
+// RandomizeOutputPosition randomizes the position of a transaction's output by
+// swapping it with a random output.  The new index is returned.  This should be
+// done before signing.
+func RandomizeOutputPosition(outputs []*wire.TxOut, index int) int {
+	r := cprng.Int31n(int32(len(outputs)))
+	outputs[r], outputs[index] = outputs[index], outputs[r]
+	return int(r)
+}
+
+// RandomizeChangePosition randomizes the position of an authored transaction's
+// change output.  This should be done before signing.
+func (tx *AuthoredTx) RandomizeChangePosition() {
+	tx.ChangeIndex = RandomizeOutputPosition(tx.Tx.TxOut, tx.ChangeIndex)
+}
+
+// SecretsSource provides private keys and redeem scripts necessary for
+// constructing transaction input signatures.  Secrets are looked up by the
+// corresponding Address for the previous output script.  Addresses for lookup
+// are created using the source's blockchain parameters and means a single
+// SecretsSource can only manage secrets for a single chain.
+//
+// TODO: Rewrite this interface to look up private keys and redeem scripts for
+// pubkeys, pubkey hashes, script hashes, etc. as separate interface methods.
+// This would remove the ChainParams requirement of the interface and could
+// avoid unnecessary conversions from previous output scripts to Addresses.
+// This can not be done without modifications to the txscript package.
+type SecretsSource interface {
+	txscript.KeyDB
+	txscript.ScriptDB
+	ChainParams() *chaincfg.Params
+}
+
+// AddAllInputScripts modifies transaction a transaction by adding inputs
+// scripts for each input.  Previous output scripts being redeemed by each input
+// are passed in prevPkScripts and the slice length must match the number of
+// inputs.  Private keys and redeem scripts are looked up using a SecretsSource
+// based on the previous output script.
+func AddAllInputScripts(tx *wire.MsgTx, prevPkScripts [][]byte, secrets SecretsSource) error {
```

Comment:
Any interest in including a similar function: `AddControlledInputScripts` (feel free to suggest a better name :tongue:), which would only attempt to sign an input if the wallet can discern if it is in control os said input? 

Such a function would be useful to prospective callers who wish to engage in multi-party collaborative transaction creation+signing (CoinJoin, Lightning, etc)